tests/frontend/test_devicespecs.py

```diff
@@ -0,0 +1,146 @@
+
+        assert not dummy.parameter_ranges
+
+    def test_program_topology_construction(self):
+        """If a device spec includes a Blackbird program,
+        the topology property should return the equivalent
+        directed acyclic graph.
+        """
+
+        class DummyDevice(DeviceSpecs):
+            """Dummy device used to instantiate
+            the abstract base class"""
+            modes = 0
+            remote = False
+            interactive = True
+            primitives = set()
+            decompositions = set()
+
+            blackbird_template = textwrap.dedent(
+                """\
+                name test
+                version 0.0
+
+                Sgate(0.543, 0) | 0
+                Dgate(-7.123) | 1
+                BSgate(0.54) | 0, 1
+                MeasureFock() | 0
+                MeasureFock() | 2
+                """
+            )
+
+        dummy = DummyDevice()
+        top = dummy.topology
+
+        circuit = top.nodes().data()
+
+        # check circuit is correct length
+        assert len(circuit) == 5
+
+        # check gates are correct
+        assert circuit[0]['name'] == 'Sgate'
+        assert circuit[1]['name'] == 'Dgate'
+        assert circuit[2]['name'] == 'BSgate'
+        assert circuit[3]['name'] == 'MeasureFock'
+        assert circuit[4]['name'] == 'MeasureFock'
+
+        # check topology/edges between nodes
+        edges = {(i, j) for i, j, d in top.edges().data()}
+        assert edges == {(0, 2), (1, 2), (2, 3)}
+
+    def test_template_topology_construction(self):
+        """If a device spec includes a Blackbird template,
+        the topology property should return the equivalent
+        directed acyclic graph, with all parameters set to zero.
+        """
+
+        class DummyDevice(DeviceSpecs):
+            """Dummy device used to instantiate
+            the abstract base class"""
+            modes = 0
+            remote = False
+            interactive = True
+            primitives = set()
+            decompositions = set()
+
+            blackbird_template = textwrap.dedent(
+                """\
+                name test
+                version 0.0
+
+                Sgate({sq}, 0) | 0
+                Dgate(-7.123) | 1
+                BSgate({theta}) | 0, 1
+                MeasureFock() | 0
+                MeasureFock() | 2
+                """
+            )
+
+        dummy = DummyDevice()
+        top = dummy.topology
+
+        circuit = top.nodes().data()
+
+        # check circuit is correct length
+        assert len(circuit) == 5
+
+        # check gates are correct
+        assert circuit[0]['name'] == 'Sgate'
+        assert circuit[1]['name'] == 'Dgate'
+        assert circuit[2]['name'] == 'BSgate'
+        assert circuit[3]['name'] == 'MeasureFock'
+        assert circuit[4]['name'] == 'MeasureFock'
+
+        # check arguments
+        assert circuit[0]['args'] == [0, 0]
+        assert circuit[1]['args'] == [-7.123]
```

Comment:
Just curious why the Dgate parameter does not get converted to zero?

Reply:
Because it is hardcoded into the circuit attribute:
```python
circuit = """
    name test
    version 0.0

    Sgate({sq}, 0) | 0
    Dgate(-7.123) | 1
    BSgate({theta}) | 0, 1
    MeasureFock() | 0
    MeasureFock() | 2
    """
```
This takes into account devices that potentially have fixed gate parameters, i.e., fixed squeezing.

Reply:
ok so the parameters that are programmable should all get converted to zeros, but others will stay as specified?

Reply:
Yes. I'm not sure I like this (using zero/non-zero to distinguish free/fixed parameters), but it needs more thinking.